### PR TITLE
ISPN-9873 Remove listeners when lock is removed

### PR DESCRIPTION
**Pull request by karesti**
This is causing a memory leak! 

https://issues.jboss.org/browse/ISPN-9873

### CODE REVIEW

File: lock/src/main/java/org/infinispan/lock/impl/manager/EmbeddedClusteredLockManager.java [hunk 1]
```diff
@@ -217,6 +227,12 @@ public void execute(Runnable runnable) {
       executor.execute(runnable);
    }
 
+
+   //Visible for testing
+   AdvancedCache<ClusteredLockKey, ClusteredLockValue> getCache() {
```
Comment (danberindei) [inline lock/src/main/java/org/infinispan/lock/impl/manager/EmbeddedClusteredLockManager.java]: Not sure you need this, the test could use `cache(0, ClusteredLockModuleLifecycle.CLUSTERED_LOCK_CACHE_NAME)`

File: lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java [hunk 2]
```diff
@@ -125,4 +125,11 @@ public int countScheduledRequests() {
    public ScheduledRequest get(String requestId) {
       return scheduledRequests.get(requestId);
    }
+
+   /**
+    * Clears all the scheduled requests
+    */
+   public void clearAll() {
```
Comment (danberindei) [inline lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java]: `All` is kind of implied in `clear` :)

File: lock/src/main/java/org/infinispan/lock/impl/manager/EmbeddedClusteredLockManager.java (hunk 1 above)
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/impl/manager/EmbeddedClusteredLockManager.java]: ok

File: lock/src/main/java/org/infinispan/lock/impl/lock/ClusteredLockImpl.java [hunk 3]
```diff
@@ -60,26 +60,38 @@
...
       requestExpirationScheduler = new RequestExpirationScheduler(clusteredLockManager.getScheduledExecutorService());
-      clusteredLockCache.getCacheManager().addListener(new ClusterChangeListener());
-      clusteredLockCache.addListener(new LockReleasedListener(), new ClusteredLockFilter(lockKey));
+      clusterChangeListener = new ClusterChangeListener();
+      lockReleasedListener = new LockReleasedListener();
+      this.clusteredLockCache.getCacheManager().addListener(clusterChangeListener);
+      this.clusteredLockCache.addListener(lockReleasedListener, new ClusteredLockFilter(lockKey));
+   }
+
+   public void clear() {
```
[the remark this answers is not in the document]
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/impl/lock/ClusteredLockImpl.java]: ok

File: lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java (hunk 2 above)
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java]: ok

**Comment by danberindei** [comment]
Integrated, thanks Katia!

**Pull request merged**